pkg/tsdb/calculator.go

```diff
@@ -53,16 +61,28 @@ func (i *Interval) Milliseconds() int64 {
 	return i.Value.Nanoseconds() / int64(time.Millisecond)
 }
 
-func (ic *intervalCalculator) Calculate(timerange backend.TimeRange, minInterval time.Duration) Interval {
+func (ic *intervalCalculator) Calculate(timerange backend.TimeRange, intrvl time.Duration, intervalMode IntervalMode) Interval {
 	to := timerange.To.UnixNano()
 	from := timerange.From.UnixNano()
-	intrvl := time.Duration((to - from) / defaultRes)
+	calculatedIntrvl := time.Duration((to - from) / defaultRes)
 
-	if intrvl < minInterval {
-		return Interval{Text: interval.FormatDuration(minInterval), Value: minInterval}
+	switch intervalMode {
+	case Min:
+		if calculatedIntrvl < intrvl {
+			return Interval{Text: interval.FormatDuration(intrvl), Value: intrvl}
+		}
+	case Max:
+		if calculatedIntrvl > intrvl {
+			return Interval{Text: interval.FormatDuration(intrvl), Value: intrvl}
+		}
+	case Exact:
+		return Interval{Text: interval.FormatDuration(intrvl), Value: intrvl}
+
+	default:
+		panic(fmt.Sprintf("Unrecognized intervalMode %v", intervalMode))
```

Comment:
maybe instead of panicking here we could return (Interval, error) and here we could return empty Interval struct and an error that the intervalMode is invalid

Reply:
👍 makes sense. Updated in https://github.com/grafana/grafana/pull/36796/commits/c42ebe7d3f2b71d385e26fcac48de5699df70a03